### ISSUE TITLE
dump_only=True regression in latest Flask-Rebar v2 release candidate

### ISSUE THREAD

**Issue by jab** [plangrid/flask-rebar#248]
```python
# app.py
from itertools import count
from marshmallow import Schema, fields
from flask_rebar import Rebar
from flask import Flask

rebar = Rebar()
registry = rebar.create_handler_registry(prefix="/api")


class TodoSchema(Schema):
    id = fields.Integer(dump_only=True)
    task = fields.String(required=True)


_next_id = count().__next__
_todos_db = {}


@registry.handles(
    rule="/todos/",
    method="POST",
    request_body_schema=TodoSchema(),
    response_body_schema={201: TodoSchema()},
)
def create_todo():
    task = rebar.validated_body["task"]
    todo = {"id": _next_id(), "task": task}
    _todos_db[todo["id"]] = todo
    return todo, 201


app = Flask("repro")
rebar.init_app(app)
```

```sh
> pip list | grep flask-rebar
flask-rebar  1.12.3

> flask run &
...
 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)

> curl -H'content-type: application/json' -d'{"task": "foo"}' http://127.0.0.1:5000/api/todos/
127.0.0.1 - - [27/Jul/2021 16:08:45] "POST /api/todos/ HTTP/1.1" 201 -
{"id":0,"task":"foo"}

> pip install flask-rebar==2.0.0rc4

> curl -H'content-type: application/json' -d'{"task": "foo"}' http://127.0.0.1:5000/api/todos/
[2021-07-27 16:10:34,127] ERROR in app: Exception on /api/todos/ [POST]
Traceback (most recent call last):
  File "/Users/jab/tmp/rebarv2/.venv/lib/python3.9/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/jab/tmp/rebarv2/.venv/lib/python3.9/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/Users/jab/tmp/rebarv2/.venv/lib/python3.9/site-packages/flask_rebar/rebar.py", line 164, in wrapped
    marshaled = marshal(data=data, schema=schema)
  File "/Users/jab/tmp/rebarv2/.venv/lib/python3.9/site-packages/flask_rebar/utils/request_utils.py", line 91, in marshal
    return compat.dump(schema=schema, data=data)
  File "/Users/jab/tmp/rebarv2/.venv/lib/python3.9/site-packages/flask_rebar/compat.py", line 26, in dump
    schema.load(result)
  File "/Users/jab/tmp/rebarv2/.venv/lib/python3.9/site-packages/marshmallow/schema.py", line 719, in load
    return self._do_load(
  File "/Users/jab/tmp/rebarv2/.venv/lib/python3.9/site-packages/marshmallow/schema.py", line 901, in _do_load
    raise exc
marshmallow.exceptions.ValidationError: {'id': ['Unknown field.']}
127.0.0.1 - - [27/Jul/2021 16:10:34] "POST /api/todos/ HTTP/1.1" 500 -
{"message":"Sorry, there was an internal error.","rebar_error_code":"internal_server_error"}
```

**Comment by RookieRick** [comment]
Well that's embarrassing. Thanks for catching that!  I'll update the docs to say we don't support `dump_only`...
:trollface: Kidding of course, I'm looking into this one; I initially thought "maybe we can just do a simple `schema.dump` here" but I'm chasing down some threads where this was discussed internally at great length to try and get a better feel for what we might break in that case

**Comment by RookieRick** [comment]
Side note: I think we just found a bug in `marshmallow` ... (which is extra-unfortunate in this case because I was excited to see I could just pull a list of `dump_only` fields from the `Schema` until I tried and it was an empty set 🤦 )

```import marshmallow as m

# declaring dump_only inline in field defs:
class MySchema(m.Schema):
    id = m.fields.Integer(dump_only=True)
    label = m.fields.String(missing="(none)")
    
instance = MySchema()
assert(instance.dump_only == set()) # this should be {"id"}, no????????????? whYYYYYYYY

# declaring dump_only explicitly as Meta:
class MyOtherSchema(m.Schema):
    id = m.fields.Integer()
    label = m.fields.String(missing="(none)")
    class Meta:
        dump_only=("id", )

other_instance = MyOtherSchema()
assert(other_instance.dump_only == {"id"})
```

**Comment by RookieRick** [comment]
Alrighty, not quite ready for PR yet but I went ahead and created a draft one in case anybody wants to take an early look, as my brain is frazzled and I gotta call it a day 🤣 . I'm not 100% convinced my approach for `set_validate_on_dump` is the best one so definitely open to suggestions there. https://github.com/plangrid/flask-rebar/pull/251 

On that note, I'm sneaking in some functionality to make validate on dump OPT-IN (in keeping with the spirit of Marshmallow's new approach of "thou shalt validate your data before you call dump"). In-house we talked about this being useful for testing - e.g., turn on validate on dump in your test cases, but leave it OFF in your real code (on the assumption that your test cases are sufficient to guarantee good data) which should give a nice performance boost.
